### PR TITLE
NNS1-2847: Use IcrcWalletPage in SnsWallet

### PR DESCRIPTION
**Pull request by dskloetd**
# Motivation

SnsWallet and IcrcWallet have entirely separate implementations even though they should have identical functionality.
The reason for this is that SNS universes are identified by the rootCanisterId while ICRC tokens are identified by the ledgerCanisterId. So at each level of they component hierarchy, tokens, accounts, balances, etc. are identified in a different way.
However if we convert the rootCanisterId to the SNS ledger canister ID, we should be able to reuse the IcrcWalletPage component and the entire hierarchy below it and then delete a lot of code.

Note: the new `SnsWallet.svelte` is based on `IcrcWallet.svelte` and not based on the old `SnsWallet.svelte`. So in the first commit, I made `SnsWallet.svelte` an exact copy of `IcrcWallet.svelte` so that in the next commit it's easy to see what the changes are relative to `IcrcWallet.svelte`.

Deleting code that became unused as a result will be done in separate PRs.

# Changes

1. Make `SnsWallet.svelte` a copy of `IcrcWallet.svelte` but with the ledger canister ID derived from the root canister ID instead of from the universe ID.
2. Include `<IcrcTokenAccountsModals />` in `Wallet.svelte` also for SNS wallets. Previously `SnsWallet` had `SnsTransactionModal` embedded, but `IcrcWalletPage` requires `IcrcTokenAccountsModals` to handle the transaction modal.
3. In `sns.services.ts`, put SNS tokens in the `tokensStore` keyed by ledger canister ID in addition to root canister ID. The tokens keyed by root canister will be removed in the future PR to make sure this is properly tested and anything depending on it depends on the ledger canister ID instead.

# Tests

1. Add `IcrcTokenAccountsModals` in `AccountsTest.svelte` to facilitate tests that open a transaction modal. Currently it only has `AccountsModals` which has the receive modals and the ICP transaction modal, but not the ICRC transaction modal. This wasn't needed before because SnsWallet had its own transaction modal embedded.
2. Adapt `SnsWallet.spec.ts` to how `IcrcWalletPage` works. Use ICRC mocks instead of SNS mocks and expecting slightly different behavior because the implementations of `SnsWallet` and `IcrcWallet` had already diverged.
3. Add a unit test for `Wallet.spec.ts` to make sure the transaction modal can be opened for the SNS wallet.
4. Add a unit test to test that SNS tokens are now keyed by ledger canister ID as well as root canister ID.
5. Update page objects to the new structure.
6. Support setting `ledgerCanisterId` with `setSnsProjects` because the ledger canister ID is now a lot more important to the implementation.
7. Tested manually ad https://qsgjb-riaaa-aaaaa-aaaga-cai.dskloet-ingress.devenv.dfinity.network/

# Todos

- [x] Add entry to changelog (if necessary).

### CODE REVIEW

File: frontend/src/tests/lib/pages/SnsWallet.spec.ts [hunk 1]
```diff
@@ -313,24 +323,40 @@ describe("SnsWallet", () => {
...
+      expect(await receiveModalPo.getFinishButtonPo().isPresent()).toBe(true);
 
       await receiveModalPo.clickFinish();
 
       await runResolvedPromises();
-      expect(snsLedgerApi.getSnsAccounts).toHaveBeenCalledTimes(4);
-      expect(snsIndexApi.getSnsTransactions).toHaveBeenCalledTimes(2);
+      // IcrcWalletPage does not reload the balance, only the transactions, in
+      // `reloadAccount`. Perhaps a bug?
+      // The the number of calls is still 2, rather than 4.
```
Comment (aterga) [inline frontend/src/tests/lib/pages/SnsWallet.spec.ts:348]:
```suggestion
      // The number of calls is still 2, rather than 4.
```

File: frontend/src/tests/lib/pages/SnsWallet.spec.ts [hunk 2]
```diff
@@ -313,24 +323,40 @@ describe("SnsWallet", () => {
...
+      expect(walletLedgerApi.getAccount).toHaveBeenCalledTimes(2);
+      // IcrcWalletTransactionsList has a hard coded 4 second delay before it
```
Comment (aterga) [inline frontend/src/tests/lib/pages/SnsWallet.spec.ts:350]: > IcrcWalletTransactionsList has a hard coded 4 second delay

Just curious: Why is this needed?

File: frontend/src/tests/lib/pages/SnsWallet.spec.ts [hunk 3]
```diff
@@ -228,52 +236,53 @@ describe("SnsWallet", () => {
...
     it("should render a balance with token", async () => {
-      vi.spyOn(snsLedgerApi, "getSnsAccounts").mockResolvedValue([
-        {
-          ...mockSnsMainAccount,
-          balanceUlps: 2_233_000_000n,
-        },
-      ]);
+      vi.spyOn(walletLedgerApi, "getAccount").mockResolvedValue({
+        ...mockSnsMainAccount,
+        balanceUlps: 2_233_000_000n,
```
Comment (aterga) [inline frontend/src/tests/lib/pages/SnsWallet.spec.ts:241]: NCR

Where does this magic number come from?

File: frontend/src/tests/lib/pages/SnsWallet.spec.ts [hunk 4]
```diff
@@ -79,40 +84,43 @@ describe("SnsWallet", () => {
 
   const rootCanisterId = rootCanisterIdMock;
   const rootCanisterIdText = rootCanisterId.toText();
+  const ledgerCanisterId = Principal.fromText("bw4dl-smaaa-aaaaa-qaacq-cai");
```
Comment (aterga) [inline frontend/src/tests/lib/pages/SnsWallet.spec.ts:87]: Why is this the only canister ID that needed to be hard coded?

Comment (dskloetd) [reply frontend/src/tests/lib/pages/SnsWallet.spec.ts:86]: As a general rule, I hard code values that matter to a test in that test.
If the value is only there to allow the test to test something else, then I don't mind using default values defined somewhere else.

To answer the question why is the `ledgerCanisterId` different from the `rootCanisterId`, it's because the `rootCanisterId` was already there before we established the rule. And I don't have the bandwidth to fix all the existing tests.

Ideally I would have gotten this test completely in line before this PR. Sometimes things aren't ideal.

Here is the document which has this rule among others:
https://docs.google.com/document/d/1sVXb7IXToniKdHsEnTb3KFz1BQ8IkrVNV59bEiJWIUM/edit#heading=h.qqurwk1mzeif
"If a test depends on specific values, these values should be defined in the same file"

File: frontend/src/tests/lib/pages/SnsWallet.spec.ts (hunk 3 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/SnsWallet.spec.ts:239]: It was already there before my change.
But this is fine IMO. It's scoped entirely within this small test and the only thing that matters is that it matches the "22.33" a few lines below. I think they are close enough together that that's clear.
If it was the same value, I would prefer using a constant defined at the start of this test.
But it's not the same value and `expect(await po.getWalletPageHeadingPo().getTitle()).toBe(formatTokenAmount(TokenAmountV2.fromUlps(balance, token));` would be less readable.

I think ideal in this case would be to define 2 constants on consecutive lines:
```
const balanceUlps = 2_233_000_000n;
const balanceFormatted = "22.33 OOO";
```
So it's completely clear that they represent the same things.

But again, this was there already and the PR is too big to mix in little cleanups like that.

File: frontend/src/tests/lib/pages/SnsWallet.spec.ts (hunk 1 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/SnsWallet.spec.ts:346]: Thanks. Done.

File: frontend/src/lib/pages/SnsWallet.svelte [hunk 5]
```diff
@@ -1,230 +1,73 @@
...
-
-  $: accountIdentifier, $snsProjectAccountsStore, load();
-
-  const reloadAccount = async () => {
-    try {
-      await Promise.all([
-        nonNullish($snsOnlyProjectStore)
-          ? syncSnsAccounts({ rootCanisterId: $snsOnlyProjectStore })
-          : Promise.resolve(),
-        reloadTransactions(),
```
Comment (mstrasinskis) [inline frontend/src/lib/pages/SnsWallet.svelte:123]: Can the `loadAccount` and `reloadTransactions` functions also be called in parallel in the `IcrcWalletPage` component?

File: frontend/src/tests/lib/pages/SnsWallet.spec.ts (hunk 2 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/SnsWallet.spec.ts:348]: The delay is here:
https://github.com/dfinity/nns-dapp/blob/709fcbd06d83f385f4a3722c01378e56cf5b6ff9/frontend/src/lib/components/accounts/IcrcWalletTransactionsList.svelte#L79

It's commented with
```
// We optimistically try to fetch the new transaction the user just transferred by delaying the reload of the transactions.
```
So I guess the assumption was that we only load transactions after making a transaction and because the transactions are loaded from the index canister, we hope that if we wait a bit, the transaction will have been copied from the ledger canister to the index canister.

In reality we have the hard coded delay also on initial load, which is very bad. And it's not clear to me we need it even after a transaction, especially since the SNS wallet didn't have it and the ICRC wallet does. But when I thought I could quickly fix it I got entangled in a large web of spaghetti and I haven't gotten back to it because nobody complains about it but me.

**Review by lmuntaner — APPROVED** [review]
Thanks!

File: frontend/src/lib/pages/SnsWallet.svelte (hunk 5 above)
Comment (dskloetd) [reply frontend/src/lib/pages/SnsWallet.svelte:123]: Perhaps, but `loadAccount` in `IcrcWalletPage` does not actually load the account.
It just copies the account from one store to another store.

File: frontend/src/tests/lib/pages/SnsWallet.spec.ts (hunk 4 above)
Comment (aterga) [reply frontend/src/tests/lib/pages/SnsWallet.spec.ts:86]: Sorry, I didn't realize that this was testing code.

File: frontend/src/tests/lib/pages/SnsWallet.spec.ts (hunk 3 above)
Comment (aterga) [reply frontend/src/tests/lib/pages/SnsWallet.spec.ts:239]: Makes sense. NCR = no change required (because clearly this is not in scope of your PR).

File: frontend/src/tests/lib/pages/SnsWallet.spec.ts (hunk 2 above)
Comment (aterga) [reply frontend/src/tests/lib/pages/SnsWallet.spec.ts:348]: Understood, thanks for clarifying.
